### PR TITLE
The draft PR of speech decoder (HF version)

### PR DESCRIPTION
**Pull request by cchen1436**
# What does this PR do ?

This PR draft of implements an independent speech decoder for duplex S2S model. 

**Collection**: [duplex-s2s]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage
Script path on dfw cluster: /lustre/fsw/portfolios/convai/users/cchen1/code/duplex_s2s
1)  debug.sh to get an GPU on interactive node
2) run.sh

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/duplex_s2s/modules/speech_decoder.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+                audio_embeddings.append(nn.Embedding(num_audio_tokens_per_codebook, self.speech_decoder_parms["d_model"]))
+
+            self.audio_embeddings = nn.ModuleList(audio_embeddings)
+
+    def forward(self, hidden_states, speech_mask, input_audio_tokens=None, return_raw_logits=False):
+        # Megatron LLM parallel training returns T, B, F so reshape it
+        # T, B, F = hidden_states.size()
+        hidden_states = hidden_states.transpose(0, 1).contiguous() # .reshape(B, T, F) # from [T, B, F] to [B, T, F]
+        # input cache needed due our transformer kv cache implementation expect the whole left context
+        if self.use_input_cache:
```
Comment (zhehuaichen) [inline nemo/collections/duplex_s2s/modules/speech_decoder.py:47]: iiuc we essentially try to reuse self.t5_decoder.cache for kv cache. + @Edresson  some questions:
1. do we need to set self.t5_decoder.use_cache=True or call self.t5_decoder.reset_cache() in the very beginning? otherwise will the kv cache be enabled in the beginning?
2. since there is cache implementation inside self.t5_decoder, why do we still need self.cache? can you help add some comments to explain here? the other way to ask this question is my understanding is the cache inside self.t5_decoder layers will handle the previous cache, so why do we still need manually doing "self.cache["hidden_states"] = torch.cat([self.cache["hidden_states"], hidden_states], dim=1)" in line 51
3. can you also explain how speech_mask interacts with kv cache?

File: nemo/collections/duplex_s2s/models/duplex_s2s_model.py [hunk 2]
```diff
@@ -633,6 +634,378 @@ def configure_model(self) -> None:
             # self.perception = checkpoint_wrapper(self.perception)
             self.perception = fully_shard(self.perception, **fsdp_config)
 
+class DuplexS2SModelSpeechDecoder(DuplexS2SModel):
+    def __init__(self, cfg) -> None:
+        super().__init__(cfg)
+        self.speech_decoder = SpeechDecoder(
```
Comment (zhehuaichen) [inline nemo/collections/duplex_s2s/models/duplex_s2s_model.py:640]: how about calling it self.speech_generation

File: nemo/collections/duplex_s2s/models/duplex_s2s_model.py [hunk 3]
```diff
@@ -633,6 +634,378 @@ def configure_model(self) -> None:
...
+        audio_logits = audio_logits.view(B, T, self._codebook_size, self._num_codebooks)
+
+
+        return {
+            "text_logits": text_logits,
+            "audio_logits": audio_logits,
+        }
+
+
+    def prepare_inputs(self, batch: dict):
```
Comment (zhehuaichen) [inline nemo/collections/duplex_s2s/models/duplex_s2s_model.py:674]: rmb to diff with the latest prepare_inputs and training_step from Piotr and bring over necessary changes

File: nemo/collections/duplex_s2s/models/duplex_s2s_model.py [hunk 4]
```diff
@@ -633,6 +634,378 @@ def configure_model(self) -> None:
...
+            #     self.tokenizer.ids_to_text(hyp_ids) for hyp_ids in forward_outputs["text_logits"].argmax(dim=-1).cpu()
+            # ]
+            # ref = [[tt] for tt in dataset_batch["target_texts"]]
+            # for h, (r,) in zip(hyp, ref):
+            #     print(f"[TEXT] Ref: {r}\n[TEXT] Hyp: {h}")
+            # # Text BLEU
+            # self.text_bleu[name].update(hyp, ref)
+
+    @property
+    def oomptimizer_schema(self) -> dict:
```
Comment (zhehuaichen) [inline nemo/collections/duplex_s2s/models/duplex_s2s_model.py:882]: remove this?

File: nemo/collections/duplex_s2s/models/duplex_s2s_model.py [hunk 5]
```diff
@@ -633,6 +634,378 @@ def configure_model(self) -> None:
...
+            self.lm_head = fully_shard(self.lm_head, **fsdp_config)
+            # self.audio_head = checkpoint_wrapper(self.audio_head)
+            self.audio_head = fully_shard(self.audio_head, **fsdp_config)
+
+            # for idx, layer in enumerate(self.perception.encoder.layers):
+            #     self.perception.encoder.layers[idx] = fully_shard(layer, **fsdp_config)
+            # self.perception = checkpoint_wrapper(self.perception)
+            self.perception = fully_shard(self.perception, **fsdp_config)
+
+    def configure_optimizers(self):
```
Comment (zhehuaichen) [inline nemo/collections/duplex_s2s/models/duplex_s2s_model.py:992]: document what's changed compared to base class

File: nemo/collections/duplex_s2s/models/duplex_s2s_model.py [hunk 6]
```diff
@@ -633,6 +634,378 @@ def configure_model(self) -> None:
...
+                # layer.self_attn = checkpoint_wrapper(layer.self_attn)
+                # layer.mlp = checkpoint_wrapper(layer.mlp)
+                self.llm.layers[idx] = fully_shard(layer, **fsdp_config)
+            self.embed_tokens = fully_shard(self.embed_tokens, **fsdp_config)
+            # self.embed_audio_tokens = fully_shard(self.embed_audio_tokens, **fsdp_config)
+            self.llm = fully_shard(self.llm, **fsdp_config)
+            # self.lm_head = checkpoint_wrapper(self.lm_head)
+            self.lm_head = fully_shard(self.lm_head, **fsdp_config)
+            # self.audio_head = checkpoint_wrapper(self.audio_head)
+            self.audio_head = fully_shard(self.audio_head, **fsdp_config)
```
Comment (zhehuaichen) [inline nemo/collections/duplex_s2s/models/duplex_s2s_model.py:985]: remove this?

File: nemo/collections/duplex_s2s/models/duplex_s2s_model.py [hunk 7]
```diff
@@ -633,6 +634,378 @@ def configure_model(self) -> None:
...
+
+            # for idx, layer in enumerate(self.perception.encoder.layers):
+            #     self.perception.encoder.layers[idx] = fully_shard(layer, **fsdp_config)
```
Comment (zhehuaichen) [inline nemo/collections/duplex_s2s/models/duplex_s2s_model.py:988]: add self.speech_generation here?

File: nemo/collections/duplex_s2s/models/duplex_s2s_model.py [hunk 8]
```diff
@@ -633,6 +634,378 @@ def configure_model(self) -> None:
...
+                {
+                    "name": "target_tokens",
+                    "type": NeuralType(("B", "T"), LabelsType()),
+                    "seq_length": "output",
+                    "vocab_size": self.tokenizer.vocab_size,
+                },
+            ],
+        }
+
+    def configure_model(self) -> None:
```
Comment (zhehuaichen) [inline nemo/collections/duplex_s2s/models/duplex_s2s_model.py:903]: document what's changed compared to base class

**Review by zhehuaichen — COMMENTED** [review]
Thanks for the work!

File: nemo/collections/tts/modules/t5tts_transformer.py [hunk 9]
```diff
@@ -0,0 +1,710 @@
...
+
+import torch
+import torch.nn.functional as F
+from nemo.utils import logging
+
+# TODO: Move the cache implementation out of the Module class, and pass it as part of the forward so we can reset
+# as needed in the inference pipeline.
+
+
+class ConvolutionLayer(torch.nn.Module):
```
[the remark this answers is not in the document]
Comment (Edresson) [reply nemo/collections/tts/modules/t5tts_transformer.py:26]: Yes, we can reuse transformer_2501.py classes. I was planing to do it soon we merge with the main branch.

File: nemo/collections/tts/models/audio_codec.py [hunk 10]
```diff
@@ -32,6 +32,7 @@
     SISDRLoss,
     TimeDomainLoss,
 )
+from nemo.collections.tts.modules.audio_codec_modules import PhonemeASR, ResNetSpeakerEncoder
```
[the remark this answers is not in the document]
Comment (Edresson) [reply nemo/collections/tts/models/audio_codec.py:35]: Audio Codec changes was done to make the model support the Causal architecture and also to make it support bf16-mixed training/inference.

File: nemo/collections/tts/modules/t5tts_transformer.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (zhehuaichen) [reply nemo/collections/tts/modules/t5tts_transformer.py:26]: @cchen1436 can you try to do that and discuss with Edresson if any issues

File: nemo/collections/duplex_s2s/modules/speech_decoder.py [hunk 11]
```diff
@@ -0,0 +1,186 @@
+import torch
+from omegaconf import DictConfig
+from torch import nn
+
+from nemo.core.classes.module import NeuralModule
+from nemo.collections.tts.modules import t5tts_transformer
+
+class SpeechDecoder(NeuralModule):
```
[the remark this answers is not in the document]
Comment (Edresson) [reply nemo/collections/duplex_s2s/modules/speech_decoder.py:8]: Perhaps TransformerSpeechDecoder would be a better name? That way, we can define other specialized classes like FlowSpeechDecoder, and so on. TransformerDecoder feels too generic and doesn't clearly indicate that its output will always be speech or audio tokens.

File: nemo/collections/duplex_s2s/modules/speech_decoder.py [hunk 12]
```diff
@@ -0,0 +1,186 @@
+import torch
```
[the remark this answers is not in the document]
Comment (Edresson) [reply nemo/collections/duplex_s2s/modules/speech_decoder.py:1]: I'm fine with this change, make sense.

File: examples/duplex_s2s/s2s_duplex_train.py [hunk 13]
```diff
@@ -19,7 +19,7 @@
 from omegaconf import OmegaConf
 
 from nemo.collections.duplex_s2s.data.datamodule import S2SDataModule
-from nemo.collections.duplex_s2s.models.duplex_s2s_model import DuplexS2SModel
+from nemo.collections.duplex_s2s.models.duplex_s2s_model import DuplexS2SModel, DuplexS2SModelSpeechDecoder
```
Comment (Edresson) [inline examples/duplex_s2s/s2s_duplex_train.py:22]: I think we should remove the DuplexS2SModel import here.

File: nemo/collections/duplex_s2s/modules/speech_decoder.py (hunk 1 above)
Comment (zhehuaichen) [reply nemo/collections/duplex_s2s/modules/speech_decoder.py:47]: Can you remind me which part of the current code is related to `our current transformer implementation expects the whole input` for caching stuffs?
